pkg/controller/acmechallenges/update.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package acmechallenges
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	internalchallenges "github.com/cert-manager/cert-manager/internal/controller/challenges"
+	"github.com/cert-manager/cert-manager/internal/controller/feature"
+	cmacme "github.com/cert-manager/cert-manager/pkg/apis/acme/v1"
+	"github.com/cert-manager/cert-manager/pkg/client/clientset/versioned"
+	utilfeature "github.com/cert-manager/cert-manager/pkg/util/feature"
+	"github.com/cert-manager/cert-manager/test/unit/gen"
+)
+
+type updater interface {
+	update(context.Context, *cmacme.Challenge) (*cmacme.Challenge, error)
+	updateStatus(context.Context, *cmacme.Challenge) (*cmacme.Challenge, error)
+}
+
+type objectUpdater struct {
+	updater
+}
+
+func newObjectUpdater(cl versioned.Interface, fieldManager string) *objectUpdater {
+	o := &objectUpdater{
+		updater: &objectUpdaterDefault{cl: cl},
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(feature.ServerSideApply) {
+		o.updater = &objectUpdaterSSA{
+			fieldManager: fieldManager,
+			cl:           cl,
+		}
+	}
+	return o
+}
+
+// updateObject updates the Finalizers if they have changed and updates the Status if it has changed.
+// Finalizers are updated using the Update method while Status is updated using
+// the UpdateStatus method.
+// Both updates will be attempted, even if one fails, except in the case where
+// one of the updates fails with a Not Found error.
+// If the any of the API operations results in a Not Found error, updateObject
+// will exit without error and the remaining operations will be skipped.
+// Only the Finalizers and Status fields may be modified. If there are any
+// modifications to new object, outside of the Finalizers and Status fields,
+// this function will panic.
+func (o *objectUpdater) updateObject(ctx context.Context, old, new *cmacme.Challenge) error {
+	if !apiequality.Semantic.DeepEqual(
+		gen.ChallengeFrom(old, gen.SetChallengeFinalizers(nil), gen.ResetChallengeStatus()),
+		gen.ChallengeFrom(new, gen.SetChallengeFinalizers(nil), gen.ResetChallengeStatus()),
+	) {
+		panic("only the finalizers and status fields may be modified")
```

Comment:
Suggestion: although this should never happen, I think we should not panic as we don't want to crash cert-manager just because there is one `Challenge` in a weird state. Perhaps we could just set it to Errored if we think it is not possible to recover from this state?

Reply:
Done. I've changed it to an `argumentError` and I catch that, log it, and stop the challenge being retried in that case. But I do think we should talk about what consitutes a programming error. 
If our tests are any good, this panic would have crashed a test long before it gets shipped.

Reply:
I agree that it would be interesting to have a discussion about this- I also have at times wondered what is a good approach in cases of errors that users should never see. But I guess it would still be better to show them an error they should never see than crashing the process.

Reply:
It needn't crash the process.

There are plenty of examples of this in Kubernetes, here for example:
 * https://github.com/kubernetes-sigs/structured-merge-diff/search?q=panic

And the sample-controller uses [runtimeutil.HandleCrash 
](https://github.com/kubernetes/apimachinery/blob/15885e270e2cefaaff9aa2b7b05eb004192bb5db/pkg/util/runtime/runtime.go#L40-L59)
at various points to avoid a panic in a Sync function crashing the worker thread:

 * https://github.com/kubernetes/apimachinery/blob/15885e270e2cefaaff9aa2b7b05eb004192bb5db/pkg/util/runtime/runtime.go#L40-L59
 * And via [wait.Until](https://github.com/kubernetes/kubernetes/blob/22a21f974f5c0798a611987405135ab7e62502da/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L153-L156)

So perhaps we should be doing the same in cert-manager.